TFEngine.py

```diff
@@ -1857,86 +1857,127 @@ def search(self, dataset, do_eval=True, output_layer_name="output", output_file=
-        if target_key and dataset.can_serialize_data(target_key):
-          print("  ref:", dataset.serialize_data(key=target_key, data=targets[i]), file=log.v4)
-          if out_beam_size is None:
-            print("  hyp:", dataset.serialize_data(key=target_key, data=output[out_idx]), file=log.v4)
-          else:
-            assert beam_scores is not None
-            for b in range(out_beam_size):
-              print(
-                "  hyp %i, score %f:" % (b, beam_scores[i][b]),
-                dataset.serialize_data(key=target_key, data=output[out_idx + b]),
-                file=log.v4)
         if out_cache is not None:
           corpus_seq_idx = dataset.get_corpus_seq_idx(seq_idx[i])
           assert corpus_seq_idx not in out_cache
           seq_idx_to_tag[corpus_seq_idx] = seq_tag[i]
-          if out_beam_size is None:
-            out_cache[corpus_seq_idx] = dataset.serialize_data(key=target_key, data=output[out_idx])
+          if num_serializable_targets > 1:
+            out_cache[corpus_seq_idx] = {}
+
+        for t in range(num_targets):
+          if out_beam_sizes[t] is None:
+            print("seq_idx: %i, seq_tag: %r, output %r: %r" % (seq_idx[i], seq_tag[i], target_keys[t], outputs[t][i]), file=log.v4)
+            out_idx = i
           else:
-            assert beam_scores is not None
-            out_cache[corpus_seq_idx] = [
-              (beam_scores[i][b], dataset.serialize_data(key=target_key, data=output[out_idx + b]))
-              for b in range(out_beam_size)]
+            print("seq_idx: %i, seq_tag: %r, outputs %r: %r" % (
+              seq_idx[i], seq_tag[i], output_layer_names[t], outputs[t][i * out_beam_sizes[t]:(i + 1)*out_beam_sizes[t]]), file=log.v4)
+            out_idx = i * out_beam_sizes[t]
+          if target_keys[t] and dataset.can_serialize_data(target_keys[t]):
+            print("  ref:", dataset.serialize_data(key=target_keys[t], data=targets[t][i]), file=log.v4)
+            if out_beam_sizes[t] is None:
+              print("  hyp:", dataset.serialize_data(key=target_keys[t], data=outputs[t][out_idx]), file=log.v4)
+            else:
+              assert beam_scores[t] is not None
+              for b in range(out_beam_sizes[t]):
+                print(
+                  "  hyp %i, score %f:" % (b, beam_scores[t][i][b]),
+                  dataset.serialize_data(key=target_keys[t], data=outputs[t][out_idx + b]),
+                  file=log.v4)
+
+            if out_cache is not None:
+              if out_beam_sizes[t] is None:
+                  out_data = dataset.serialize_data(key=target_keys[t], data=outputs[t][out_idx])
+              else:
+                assert beam_scores[t] is not None
+                out_data = [
+                    (beam_scores[t][i][b], dataset.serialize_data(key=target_keys[t], data=outputs[t][out_idx + b]))
+                    for b in range(out_beam_sizes[t])]
+
+              if num_serializable_targets == 1:
```

Comment:
The check should be `isinstance(output_layer_names, dict)`, and not `num_serializable_targets == 1`.

Reply:
list you mean? But I convert output_layer_names to list if it is a string, should I change that?

Reply:
Sorry, yes, I meant list. You can also do `want_output_list = isinstance(output_layer_names, list)` in the beginning. Or do not convert `output_layer_names`. Whatever makes the code simpler.